internal/certprovider.go

```diff
@@ -1,59 +1,174 @@
 
 import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha256"
 	"crypto/x509"
 	_ "embed"
+	"encoding/hex"
 	"encoding/pem"
 	"fmt"
+	"io"
+	"net/http"
+	"strings"
 )
 
-// asserts/aws_nitro_enclaves.crt contains the PEM encoded roots for verifying Nitro
-//	Enclave attestation signatures. You can download them from
-//	https://aws-nitro-enclaves.amazonaws.com/AWS_NitroEnclaves_Root-G1.zip
-//	It's recommended you calculate the SHA256 sum of this string and match
-//	it to the one supplied in the AWS documentation
+// AWS_NitroEnclaves_Root-G1.zip file and its digest are advertised on
 //	https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html
 
-//go:embed assets/aws_nitro_enclaves.crt
-var NitroCertPEM []byte
+const AWSNitroEnclavesRootURL = "https://aws-nitro-enclaves.amazonaws.com/AWS_NitroEnclaves_Root-G1.zip"
 
-//go:embed assets/selfsigned_cert.der
-var selfSignedCertDER []byte
+//go:embed assets/AWS_NitroEnclaves_Root-G1.zip
+var AWSNitroEnclavesRootZip []byte
+
+//go:embed assets/AWS_NitroEnclaves_Root-G1.sha256.hex
+var AWSNitroEnclavesRootSHA256Hex string
+
+type UnzipAWSRootCertsFunc func(zipBytes []byte) (pemBytes []byte, err error)
+
+func UnzipAWSRootCerts(zipBytes []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(
+		bytes.NewReader(zipBytes),
+		int64(len(zipBytes)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("creating zip reader: %w", err)
+	}
+
+	if len(zipReader.File) != 1 {
+		return nil, fmt.Errorf("unexpected file count: %d", len(zipReader.File))
+	}
+	if zipReader.File[0].Name != "root.pem" {
+		return nil, fmt.Errorf("unexpected file name: %s", zipReader.File[0].Name)
+	}
+
+	pemFile, err := zipReader.File[0].Open()
+	if err != nil {
+		return nil, fmt.Errorf("opening zip file: %w", err)
+	}
+
+	pemBytes, err := io.ReadAll(pemFile)
+	if err != nil {
+		return nil, fmt.Errorf("reading pem file: %w", err)
+	}
+
+	return pemBytes, nil
+}
+
+type ExtractRootsFunc func(
+	rootsZIPBytes []byte,
+	rootsDigestHex string,
+	unzipAWSRootCerts UnzipAWSRootCertsFunc,
+) (
+	*x509.CertPool,
+	error,
+)
+
+func ExtractRoots(
+	rootsZIPBytes []byte,
+	rootsDigestHex string,
+	unzipAWSRootCerts UnzipAWSRootCertsFunc,
+) (
+	*x509.CertPool,
+	error,
+) {
+	digest := sha256.Sum256(rootsZIPBytes)
+	digestHex := hex.EncodeToString(digest[:])
+	if digestHex != strings.TrimSpace(rootsDigestHex) {
+		return nil,
+			fmt.Errorf("digest mismatch: %s != %s", digestHex, rootsDigestHex)
+	}
+
+	rootCertsPEM, err := unzipAWSRootCerts(rootsZIPBytes)
+	if err != nil {
+		return nil, fmt.Errorf("unzipping roots: %w", err)
+	}
+
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(rootCertsPEM)
+	if !ok {
+		return nil, fmt.Errorf("appending cert")
+	}
+	return roots, nil
+}
 
 type NitroCertProvider struct {
-	certs *x509.CertPool
+	RootCerts    *x509.CertPool
+	ExtractRoots ExtractRootsFunc
 }
 
 func NewNitroCertProvider() *NitroCertProvider {
-	return &NitroCertProvider{}
+	return &NitroCertProvider{
+		RootCerts:    nil,
```

Comment:
The `RootsCert` variable is exported only to facilitate tests. You're right though that it was leaking in `nitrite.go`. The plan is to eventually export much less out of the `nitrite` module and also to build a `nitrite.Verifier` with options that will hide a lot of the internals.

In the meantime, I'm avoiding the leakage by functions in `nitrite.go` that provide different kinds of `CertProviders`
```go
type CertProvider interface {
	Roots() (*x509.CertPool, error)
}

func NewNitroCertProvider() CertProvider {
	return internal.NewNitroCertProvider(
		internal.NewEmbeddedRootCertZipReader(),
	)
}

func NewFetchingNitroCertProvider() (CertProvider, error) {
	reader, err := internal.NewFetchingRootCertZipReader()
	if nil != err {
		return nil, fmt.Errorf("creating fetching root cert zip reader: %w", err)
	}

	return internal.NewNitroCertProvider(reader), nil
}

func NewSelfSignedCertProvider() CertProvider {
	return internal.NewSelfSignedCertProvider()
}
```
The usual practice is to return structs, but I'm wonder if in this case returning interfaces is actually a nice mechanism to restrict the access to the internals of each `CertProvider` struct.

Reply:
OK, I changed the internal implementation so taht `RootsCert` is not exported.
I'm still caching `rootCert` (as opposed to zipBytes) since that's already implemented, but I added a comment that we do that on the account of the ReadCloser being closed after the first call.
I'm still returning the interfaces, though that whole thing will likely change once we redefine the options for `nitrite.Verifier`.